File: src/main/java/me/itzisonn_/meazy/parser/pasing_function/expression/MultiplicationExpressionParsingFunction.java
package me.itzisonn_.meazy.parser.pasing_function.expression;

import me.itzisonn_.meazy.MeazyMain;
import me.itzisonn_.meazy.lexer.TokenTypeSets;
import me.itzisonn_.meazy.parser.ParsingContext;
import me.itzisonn_.meazy.parser.Parser;
import me.itzisonn_.meazy.parser.ast.expression.Expression;
import me.itzisonn_.meazy.parser.operator.OperatorType;
import me.itzisonn_.meazy.parser.ast.expression.OperatorExpression;
import me.itzisonn_.meazy.parser.pasing_function.AbstractParsingFunction;
import org.jspecify.annotations.NullMarked;
import org.jspecify.annotations.Nullable;

@NullMarked
public class MultiplicationExpressionParsingFunction extends AbstractParsingFunction<Expression> {
    public MultiplicationExpressionParsingFunction() {
        super("multiplication_expression");
    }

    @Override
    public Expression parse(ParsingContext context, @Nullable Object... extra) {
        Parser parser = context.getParser();
        Expression left = parser.parseAfter(MeazyMain.getDefaultIdentifier("multiplication_expression"), Expression.class);

        while (TokenTypeSets.INSTANCE.getMultiplication().contains(parser.getCurrent().getType())) {
            String operator = parser.getCurrentAndNext().getValue();
            Expression right = parser.parseAfter(MeazyMain.getDefaultIdentifier("multiplication_expression"), Expression.class);
            left = new OperatorExpression(left, right, operator, OperatorType.INFIX);
        }

        return left;
    }
}

File: src/main/java/me/itzisonn_/meazy/parser/pasing_function/expression/MultiplicationExpressionParsingFunction.kt
package me.itzisonn_.meazy.parser.pasing_function.expression

import me.itzisonn_.meazy.MeazyMain.getDefaultIdentifier
import me.itzisonn_.meazy.lexer.TokenTypeSets.multiplication
import me.itzisonn_.meazy.parser.ParsingContext
import me.itzisonn_.meazy.parser.ast.expression.Expression
import me.itzisonn_.meazy.parser.ast.expression.OperatorExpression
import me.itzisonn_.meazy.parser.operator.OperatorType
import me.itzisonn_.meazy.parser.pasing_function.AbstractParsingFunction

object MultiplicationExpressionParsingFunction : AbstractParsingFunction<Expression>("multiplication_expression") {
    override fun parse(context: ParsingContext, vararg extra: Any?): Expression {
        val parser = context.parser
        var left = parser.parseAfter<Expression>(getDefaultIdentifier("multiplication_expression"))

        while (parser.current.type in multiplication) {
            val operator = parser.consume().value
            val right = parser.parseAfter<Expression>(getDefaultIdentifier("multiplication_expression"))
            left = OperatorExpression(left, right, operator, OperatorType.INFIX)
        }

        return left
    }
}
